onnxruntime/core/optimizer/matmul_transpose_fusion.cc

```diff
@@ -86,6 +86,60 @@ static size_t UpdateConsumerCount(Graph& graph, NodeArg* target, std::unordered_
   }
 }
 
+static Node* GetTransposeNodeFromCast(Graph& graph, Node* cast) {
+  /* Interchange Cast and Transpose nodes in the graph and return Transpose node if possible
+  *  Requirements to interchange Cast and Transpose nodes changing the order of the operations.
+  *  1. Both Cast and Transpose are one-output nodes (assuming both have one-input only)
+  *  2. Transpose only feeds the Cast node (and no other node)
+  *  3. Cast only feeds the MalMul node (and no other node)
+  */
+
+  ORT_ENFORCE(cast != nullptr);
+  auto transpose = GetTransposeNodeFromOutput(graph, *cast->MutableInputDefs()[0]);
+  if (transpose == nullptr || cast->GetOutputEdgesCount() !=1 || transpose->GetOutputEdgesCount() !=1) {
+    return nullptr;
+  }
+  NodeArg* cast_output = cast->MutableOutputDefs()[0];
+  NodeArg* transpose_input = transpose->MutableInputDefs()[0];
+
+  // Create a new NodeArg to feed the output from the new Cast to the new Transpose.
+  // The shape of the new NodeArg is same as the original input to Transport but type
+  // should match that of the output from the original Cast.
+
+  auto new_cast_output_type_proto = *transpose_input->TypeAsProto();
+  const ONNX_NAMESPACE::TensorProto_DataType element_type =
+      static_cast<ONNX_NAMESPACE::TensorProto_DataType>(cast_output->TypeAsProto()->tensor_type().elem_type());
+  new_cast_output_type_proto.mutable_tensor_type()->set_elem_type(element_type);
+  auto& new_cast_output = graph.GetOrCreateNodeArg(cast_output->Name() + "_transformed", &new_cast_output_type_proto);
+
+  const std::vector<NodeArg*> new_cast_input_defs {transpose_input};
+  const std::vector<NodeArg*> new_cast_output_defs {&new_cast_output};
+  const std::vector<NodeArg*> new_transpose_input_defs = {&new_cast_output};
+  const std::vector<NodeArg*> new_transpose_output_defs = {cast_output};
+
+  Node& new_cast = graph.AddNode(graph.GenerateNodeName(cast->Name() + "_transformed"),
```

Comment:
>_transformed [](start = 72, length = 12)

no strong opinion, I wonder if "_fused" is better - same for the one below

Reply:
Actually the Cast node is not fused. It is only interchanged with the Transform node

Reply:
The Transpose operation will be fused by the caller and this node anyway will be replaced by FusedMatMul. This function is only interchanging the order of the Transpose and Cast operation.